### DISCUSSION TITLE
How to use Decimal as type in fastapi?

### DISCUSSION THREAD

**Discussion by hhertogh** [fastapi/fastapi#8389] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from decimal import Decimal, getcontext

getcontext().prec = 4

class DiscountItem(BaseModel):
    name: Optional[str] = None
    total: Decimal
```

### Description

I would like to know how the precision value of the Decimal type can be used with fastapi.
I would like to have maximum 4 decimals digits but setting the precision with the getcontext method does not seem to work.
It seem this value is not taken into account by the model, is it possible ?
What can be done to make it work ?

### Operating System

macOS

### FastAPI Version

0.75.2

### Python Version

3.10.6

**Comment by JarroVGIT** [comment]
Do you want to have **validation** that the `total` always has 4 decimal digits? Or do you always want to **return** (whatever the input) a decimal with 4 decimal digits?

The first one can be achieved by using `condecimal` from Pydantic like so:
```python
from pydantic import BaseModel, condecimal

class DiscountItem(BaseModel):
    name: Optional[str] = None
    total: condecimal(decimal_places=4)
```

With regard to the latter (transform any given input value to a Decimal with precision of 4), you could use a pre-validator to transform the value. I find `Decimal` always a complicated beast. For example, when constructing a new Decimal (which is what is happening when you create a new `DiscountItem` object) the `prec` of the current context is not used. That value of `prec` is only used when performing arithmetic  operations. Using a pre-validator in your model, you might want to do something like this:

```python
class DiscountItem2(BaseModel):
    name: Optional[str] = None
    total: Decimal

    @validator("total", pre=True)
    def make_four_digits(cls, v):
        getcontext().prec = 4
        v = Decimal(value=v) * Decimal(1)
        return v
```
This will first run the `@validator` with the value of Total, before performing the real validation (e.g. parsing it to `Decimal`). By performing the arithmetic operation `Decimal(value=v) * Decimal(1)`, the `prec` is used in determining the final Decimal. 

Hope this helps!

**Comment by dotX12** [comment]
@hhertogh 

use:
```python3
percent_agent: Optional[ImprovedDecimal(round_to=0.001)] = Field(None, alias="percent")
```

custom type:

```python3
from decimal import Decimal
from typing import Any
from typing import Callable
from typing import Dict
from typing import Type
from typing import Union

from pydantic.fields import ModelField
from pydantic.types import ConstrainedDecimal  # noqa
from pydantic.types import ConstrainedFloat
from pydantic.types import ConstrainedInt
from pydantic.types import ConstrainedNumberMeta # noqa
from pydantic.types import OptionalIntFloatDecimal # noqa
from pydantic.utils import update_not_none
from pydantic.validators import decimal_validator
from pydantic.validators import number_multiple_validator
from pydantic.validators import number_size_validator

Number = Union[int, float, Decimal]
ConstrainedNumber = Union[ConstrainedDecimal, ConstrainedFloat, ConstrainedInt]


class CustomConstrainedDecimal(ConstrainedDecimal, metaclass=ConstrainedNumberMeta):
    round_to: OptionalIntFloatDecimal = None

    @classmethod
    def __modify_schema__(cls, field_schema: Dict[str, Any]) -> None:
        update_not_none(
            field_schema,
            exclusiveMinimum=cls.gt,
            exclusiveMaximum=cls.lt,
            minimum=cls.ge,
            maximum=cls.le,
            multipleOf=cls.multiple_of,
            round_to=cls.round_to,
        )

    @classmethod
    def __get_validators__(cls) -> Callable:
        yield decimal_validator
        yield number_size_validator
        yield number_multiple_validator
        yield cls.round_to_validator
        yield cls.validate

    @staticmethod
    def round_to_validator(v: "Number", field: "ModelField") -> "Number":
        field_type: CustomConstrainedDecimal = field.type_
        if field_type.round_to:
            if isinstance(field_type.round_to, (int, float)):
                TWO_PLACES = Decimal(str(field_type.round_to))
                return Decimal(v.quantize(TWO_PLACES))
            raise Exception(
                "round_to is not a int | float. Value: '{0}'".format(
                    field_type.round_to
                )
            )
        return v


class ImprovedDecimal:
    def __new__(
        cls,
        *,
        gt: Decimal = None,
        ge: Decimal = None,
        lt: Decimal = None,
        le: Decimal = None,
        max_digits: int = None,
        decimal_places: int = None,
        multiple_of: Decimal = None,
        round_to: float = None,
    ) -> Type[Decimal]:

        namespace = dict(
            gt=gt,
            ge=ge,
            lt=lt,
            le=le,
            max_digits=max_digits,
            decimal_places=decimal_places,
            multiple_of=multiple_of,
            round_to=round_to,
        )
        return type(
            "ConstrainedDecimalValue", (CustomConstrainedDecimal,), namespace
        )  # noqa
```

**Comment by ivergara** [comment]
What you need is a slightly different validator

```python
@validator("total", pre=True)
def quantize_four_decimal_places(cls, v: Any):
    return Decimal(v).quantize(Decimal('0.0001'))
```

Use the `[quantize](https://docs.python.org/3/library/decimal.html#decimal.Decimal.quantize)` method to indicate how many decimal digits you want. 

Now, you might want to define your class with a custom encoder for `Decimal` because by default it'll be casted to `float` in the response (quite unfortunate default, luckily it can be changed), which will eat up any trailing `0` that might be relevant.

```python
class DiscountItem(BaseModel):
    name: Optional[str] = None
    total: Decimal

    class Config:
        json_encoders = {
            Decimal: str
        }
````
